tools/cli/commands/connect.py

```diff
@@ -0,0 +1,263 @@
+    browser_group.add_argument(
+        '--no-launch-browser',
+        dest='launch_browser',
+        action='store_false',
+        help='do not open a browser connected to Datalab')
+    browser_group.add_argument(
+        '--launch-browser',
+        dest='launch_browser',
+        action='store_true',
+        default=True,
+        help='open a browser connected to Datalab')
+
+    return
+
+
+def connect(args, gcloud_compute):
+    """Create a persistent connection to a Datalab instance.
+
+    Args:
+      args: The Namespace object constructed by argparse
+      gcloud_compute: A function that can be called to invoke `gcloud compute`
+    """
+    instance = args.instance
+    print('Connecting to {0}'.format(instance))
+
+    datalab_address = 'http://localhost:{0}/'.format(str(args.port))
+
+    def create_tunnel():
+        """Create an SSH tunnel to the Datalab instance.
+
+        This method blocks for as long as the connection is open.
+
+        Raises:
+          KeyboardInterrupt: When the end user kills the connection
+          subprocess.CalledProcessError: If the connection dies on its own
+        """
+        cmd = ['ssh']
+        if args.quiet:
+            cmd.append('--quiet')
+        if args.zone:
+            cmd.extend(['--zone', args.zone])
+        port_mapping = 'localhost:' + str(args.port) + ':localhost:8080'
+        cmd.extend([
+            '--ssh-flag=-4',
+            '--ssh-flag=-o',
+            '--ssh-flag=LogLevel=error',
+            '--ssh-flag=-N',
+            '--ssh-flag=-L',
+            '--ssh-flag=' + port_mapping])
+        cmd.append('datalab@{0}'.format(instance))
+        with tempfile.TemporaryFile() as tf:
+            gcloud_compute(args, cmd, stdout=tf, stderr=tf)
+        return
+
+    def on_ready():
+        """Callback that handles a successful connection."""
+        print('You can now connect to Datalab at ' + datalab_address)
+        if args.launch_browser:
+            try:
+                webbrowser.open(datalab_address)
+            except webbrowser.Error as e:
+                print('Unable to open the webbrowser: ' + str(e))
+        return
+
+    def health_check(cancelled_event):
+        """Check if the Datalab instance is reachable via the connection.
+
+        After the instance is reachable, the `on_ready` method is called.
+
+        This method is meant to be suitable for running in a separate thread,
+        and takes an event argument to indicate when that thread should exit.
+
+        Args:
+          cancelled_event: A threading.Event instance that indicates we should
+            give up on the instance becoming reachable.
+        """
+        health_url = '{0}_info/'.format(datalab_address)
+        healthy = False
+        print('Waiting for Datalab to be reachable at ' + datalab_address)
+        while not cancelled_event.is_set():
+            try:
+                health_resp = urllib2.urlopen(health_url)
```

Comment:
I've thought about it, and I don't think it would add much.

This command always keeps running until the user explicitly kills it, so I can't think of any benefit a user would get from a timeout.